src/main/drivers/rssi_softpwm.c

```diff
@@ -0,0 +1,272 @@
+            if (rawWidth < 1000 + 100) {
+                if (rawWidth < 1000)
+                    pulseWidth = 0;
+                else
+                    pulseWidth = rawWidth - 1000;
+            } else {
+                // Something went wrong
+                pulseWidth = 0; // Fail
+            }
+        } else {
+            pulseWidth = rawWidth;
+        }
+        break;
+
+    default:
+        pulseWidth = 500; // Default marker
+        break;
+    }
+
+    debug[1] = pulseWidth;
+
+    // Monitor pulseWidths for parameter decision.
+
+    if (rawWidth > tmax)
+        tmax = rawWidth;
+
+    if (rawWidth < tmin)
+        tmin = rawWidth;
+
+    debug[4] = tmax;
+}
+
+static int32_t rspFilterArray[5];
+static int rspFilterPos;
+
+bool rssiSoftPwmInit(void)
+{
+    // Nothing special, standard spell.
+    // XXX Are these done in exti.c already?
+#ifdef STM32F10X
+    // enable AFIO for EXTI support
+    RCC_APB2PeriphClockCmd(RCC_APB2Periph_AFIO, ENABLE);
+#endif
+
+#if defined(STM32F3) || defined(STM32F4)
+    /* Enable SYSCFG clock otherwise the EXTI irq handlers are not called */
+    RCC_APB2PeriphClockCmd(RCC_APB2Periph_SYSCFG, ENABLE);
+#endif
+
+    ioTag_t tag = masterConfig.rssiSoftPwmConfig.ioTag;
+    if (tag == IO_TAG_NONE)
+        return false;
+
+    rspIO = IOGetByTag(tag);
+
+    IOInit(rspIO, OWNER_RSSIPWM, 0);
+    IOConfigGPIO(rspIO, IOCFG_IN_FLOATING);
+
+    EXTIHandlerInit(&rsp_extiCallbackRec, rspExtiHandler);
+    EXTIConfig(rspIO, &rsp_extiCallbackRec, NVIC_PRIO_SOFTPWM_EXTI, EXTI_Trigger_Rising);
+
+    rxtype = masterConfig.rssiSoftPwmConfig.device;
+
+    switch(rxtype) {
+    case RXTYPE_FRSKY_TFR4:
+        pulseMin = 24;
+        pulseMax = 1024;
+        break;
+
+    default:
+        pulseMin = 1;
+        pulseMax = 999;
+        break;
+    }
+
+    rspInitialized = true;
+    rspActive = false;
+
+    for (int i = 0 ; i < 5 ; i++)
+        rspFilterArray[i] = 0;
+
+    rspFilterPos = 0;
+
+    return true;
+}
+
+// rssiSoftPwmUpdate: Scheduler task
+
+void rssiSoftPwmUpdate(uint32_t currentTime)
+{
+    UNUSED(currentTime);
+
+    uint32_t value;
+    bool longPulse = false;
+    uint8_t longValue;
+
+    if (!rspInitialized) {
+        return;
+    }
+
+    ATOMIC_BLOCK(NVIC_PRIO_SOFTPWM_EXTI) {
+        if (rspInProgress) {
+            // No falling edge has been detected for a scheduling interval.
+            // Remember the pin level and disable interrupt.
+            longValue = IORead(rspIO);
+            longPulse = true;
+            rspInProgress = false;
+            EXTIEnable(rspIO, false);
+            timerRunning = false;
+        }
+    }
+
+    if (!rspActive) {
+        // Dodge the first call.
+        rspActive = true;
+    } else {
+        if (longPulse) {
+            pulseWidth = longValue ? pulseMax : pulseMin;
+        } else {
+            rspComputePulse();
+        }
+
+        if (pulseWidth >= pulseMin && pulseWidth <= pulseMax) {
```

Comment:
this test will almost always succeed, it may be easier to sanitize value in code above